src/cpu/testers/traffic_gen/gups_gen.cc

```diff
@@ -0,0 +1,414 @@
+        temp = (temp << 1) ^ ((int64_t) temp < 0 ? POLY : 0);
+    }
+
+    for (i = 62; i >= 0; i--)
+    {
+        if ((n >> i) & 1)
+            break;
+    }
+
+    ran = 0x2;
+    while (i > 0)
+    {
+        temp = 0;
+        for (j = 0; j < 64; j++)
+            if ((ran >> j) & 1)
+            temp ^= m2[j];
+        ran = temp;
+        i -= 1;
+        if ((n >> i) & 1)
+            ran = (ran << 1) ^ ((int64_t) ran < 0 ? POLY : 0);
+    }
+
+    return ran;
+}
+
+Addr
+GUPSGen::indexToAddr (uint64_t index)
+{
+    Addr ret = index * elementSize + startAddr;
+
+    return ret;
+}
+
+PacketPtr
+GUPSGen::getReadPacket(Addr addr, unsigned int size)
+{
+    RequestPtr req = std::make_shared<Request>(addr, size,
+                                            0, requestorId);
+    // Dummy PC to have PC-based prefetchers latch on; get entropy into higher
+    // bits
+    req->setPC(((Addr)requestorId) << 2);
+
+    // Embed it in a packet
+    PacketPtr pkt = new Packet(req, MemCmd::ReadReq);
+
+    uint8_t* pkt_data = new uint8_t[req->getSize()];
+    pkt->dataDynamic(pkt_data);
+
+    return pkt;
+}
+
+PacketPtr
+GUPSGen::getWritePacket(Addr addr, unsigned int size, uint8_t *data)
+{
+    RequestPtr req = std::make_shared<Request>(addr, size, 0,
+                                               requestorId);
+    // Dummy PC to have PC-based prefetchers latch on; get entropy into higher
+    // bits
+    req->setPC(((Addr)requestorId) << 2);
+
+    // Embed it in a packet
+    PacketPtr pkt = new Packet(req, MemCmd::WriteReq);
+
+    uint8_t* pkt_data = new uint8_t[req->getSize()];
+    pkt->dataDynamic(pkt_data);
+
+    // std::fill_n(pkt_data, req->getSize(), (uint8_t)requestorId);
+    std::memcpy(pkt_data, data, req->getSize());
+
+    return pkt;
+}
+
+void
+GUPSGen::handleResponse(PacketPtr pkt)
+{
+    onTheFlyRequests--;
+    if (pkt->isWrite()) {
+        DPRINTF(GUPSGen, "handleResponse: received a write resp.\n");
+        stats.totalUpdates++;
+
+        stats.totalWrites++;
+        stats.totalBytesWritten += elementSize;
+        stats.totalWriteLat += curTick() - exitTimes[pkt->req];
+
+        exitTimes.erase(pkt->req);
+        delete pkt;
+    }
+    else {
+        DPRINTF(GUPSGen, "handleResponse: received a read resp.\n");
+
+        stats.totalReads++;
+        stats.totalBytesRead += elementSize;
+        stats.totalReadLat += curTick() - exitTimes[pkt->req];
+
+        exitTimes.erase(pkt->req);
+
+        responsePool.push(pkt);
+    }
+    if (doneReading && requestPool.empty() && onTheFlyRequests == 0){
+        stats.totalTicks = curTick();
+        exitSimLoop(name() + " has encountered the exit state and will "
+                "terminate the simulation.\n");
+        return;
+    }
+}
+
+void GUPSGen::wakeUp()
+{
+    if (!nextSendEvent.scheduled() && !requestPool.empty())
+    {
+        schedule(nextSendEvent, nextCycle());
+    }
+}
+
+void
+GUPSGen::sendNextReq() {
```

Comment:
I'm going to wait to review this until it's updated based on our conversation yesterday